Add buttons to copy error details to clipboard

johnwatkins0 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
This modifies the single AMP Validated URL screen by making it easier to copy error details to the user's clipboard in pretty-printed JSON format. Users can take action in two ways:

1. By clicking this "Copy to clipboard" button in the actions for a single error (a single JSON object is copied):

<img width="379" alt="Screen Shot 2020-10-12 at 4 57 03 PM" src="https://user-images.githubusercontent.com/9020968/95793816-0c960a80-0cac-11eb-91aa-f6620ff1d22c.png">

2. By clicking this "Copy error details to clipboard" button in the bulk actions row (an array of JSON objects is copied):

<img width="613" alt="Screen Shot 2020-10-12 at 4 58 09 PM" src="https://user-images.githubusercontent.com/9020968/95793869-2cc5c980-0cac-11eb-99ba-199587171225.png">

These buttons are to be cleaned up as part of #5431.

QA steps are in #5209 

Fixes #5209

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Add buttons to copy error details to clipboard

johnwatkins0 committed:
Move clipboard functions into own module and prettyprint JSON

johnwatkins0 committed:
Resolve failing test

johnwatkins0 committed:
Add tests for clipboard-friendly error button

johnwatkins0 commented on assets/src/amp-validation/amp-validated-url-post-edit-screen.js:
```diff
@@ -389,13 +392,10 @@ const handleBulkActions = () => {
-				row.classList.toggle( 'new', ! markingAsReviewed );
-				addBeforeUnloadPrompt();
-			}
+		getURLValidationTableRows( { checkedOnly: true } ).forEach( ( row ) => {
```
This is a tangential change. I found myself copying pasting this code (which is from another PR I worked on recently) and instead created the `getURLValidationTableRows` helper function.

pierlon commented on assets/src/amp-validation/copy-to-clipboard-buttons.js:
```diff
@@ -0,0 +1,61 @@
+/**
```
Alternatively you could use [clipboard.js](https://clipboardjs.com/) (which comes bundled with WordPress), but this approach is also fine :+1:.

johnwatkins0 replied:
Good suggestion. I didn't think of this. I've refactored out the wheel-reinvention in 61e571a47

pierlon commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -1619,6 +1619,16 @@ public static function filter_tag_row_actions( $actions, WP_Term $tag ) {
 				esc_html__( 'Details', 'amp' )
 			);
+
+			$json           = json_decode( $term->description, true );
```
So for a single error the JSON would look like:

```json
{
  "code": "DISALLOWED_TAG",
  "node_attributes": {
    "src": "https:\/\/wordpressdev.lndo.site\/content\/themes\/twentytwentyone\/assets\/js\/primary-navigation.js?ver=__normalized__",
    "id": "twenty-twenty-one-primary-navigation-script-js"
  },
  "node_name": "script",
  "node_type": 1,
  "parent_name": "body",
  "type": "js_error",
  "status": "Removed"
}
```

For the `node_type` field, I'm wondering if we should get the XML constant name from the value and use that instead. By itself, the constant value wouldn't be of much use and would require the person having to lookup the constants themselves. For reference, here's [the list](https://www.php.net/manual/en/dom.constants.php):

![image](https://user-images.githubusercontent.com/16200219/96514406-2490ff80-1253-11eb-8ff1-ee6d935f75c9.png)

pierlon replied:
We could reflect the [`XMLReader`](https://www.php.net/manual/en/class.xmlreader.php) class and obtain the constants from there

johnwatkins0 replied:
Done in fe358015b

pierlon commented on assets/src/amp-validation/copy-to-clipboard-buttons.js:
```diff
@@ -0,0 +1,61 @@
+	}
+
+	const value = getURLValidationTableRows( { checkedOnly: true } ).map( ( row ) => {
+		const copyButton = row.querySelector( '.single-url-detail-copy' );
```
I noticed that if the user changes any of the row's values (without updating the post), the JSON is not updated. For example, updating a validation error to be kept and then copying the JSON for the row does not reflect the updated field value.

johnwatkins0 replied:
Fixed in c4e1a6423

johnwatkins0 committed:
Convert error node_type to constant name in JSON

johnwatkins0 committed:
Update markup status in JSON data in case it has changed

johnwatkins0 committed:
Use clipboard.js package for copying to clipboard

johnwatkins0 commented:
Thanks for the suggestions, @pierlon. I've updated and am re-requesting now.

johnwatkins0 committed:
Disable eslint rule per line

johnwatkins0 committed:
Remove unneeded helper function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=5500) <br/>All committers have signed the CLA.

pierlon commented:
The only issue I have with this currently is that the "Copy error details to clipboard" button nor the "Copy to clipboard" buttons stay focused when clicked. Also, should we provide a visual feedback to indicate that the JSON has successfully been copied?

johnwatkins0 commented:
> The only issue I have with this currently is that the "Copy error details to clipboard" button nor the "Copy to clipboard" buttons stay focused when clicked. Also, should we provide a visual feedback to indicate that the JSON has successfully been copied?

Good call, I will update.

johnwatkins0 committed:
Copy to clipboard: restore focus state and show 'copied' text

johnwatkins0 commented:
@pierlon I've addressed your suggestions in ddf5ec467

1. The button is focused on success.
2. The button text is updated for four seconds. (I referred to [this Gutenberg component](https://github.com/WordPress/gutenberg/blob/44c9ec0168e24443d19d00172cda9b2a3c64d157/packages/components/src/clipboard-button/index.js) for inspiration.)

Let me know what you think.

pierlon commented on assets/src/amp-validation/copy-to-clipboard-buttons.js:
```diff
@@ -3,28 +3,53 @@
+ * @param {HTMLElement} event.trigger The element triggering the event.
+ */
+function onSuccess( { trigger } ) {
+	trigger.focus();
```
Minor bug:
Clicking on the button while it's showing the updated text permanently sets it to said text:

![deepin-screen-recorder_Select area_20201028143033](https://user-images.githubusercontent.com/16200219/97480867-e9c25200-194b-11eb-89a4-e6ec62ee511a.gif)

johnwatkins0 replied:
Yes, looking at my code, that bug makes perfect sense 🤦 Fixing.

johnwatkins0 replied:
Thanks for the catch, @pierlon. Fixed in 47fc7be1d

johnwatkins0 committed:
Copy to clipboard: handle case where clicks button again immediately

pierlon reviewed (approved):
Well done!

westonruter commented on includes/validation/class-amp-validated-url-post-type.php:
```diff
@@ -2796,6 +2796,7 @@ public static function render_single_url_list_table( $post ) {
 			<button type="button" class="button action keep"><?php esc_html_e( 'Keep', 'amp' ); ?></button>
 			<button type="button" class="button action reviewed-toggle reviewed"><?php esc_html_e( 'Mark reviewed', 'amp' ); ?></button>
 			<button type="button" class="button action reviewed-toggle unreviewed"><?php esc_html_e( 'Mark unreviewed', 'amp' ); ?></button>
+			<button type="button" class="button action copy-all"><?php esc_html_e( 'Copy error details to clipboard', 'amp' ); ?></button>
```
I suggest making this consistent with the row action:

```suggestion
			<button type="button" class="button action copy-all"><?php esc_html_e( 'Copy to clipboard', 'amp' ); ?></button>
```

Reason is that when the button text changes from “Copy to clipboard” to “Copied to clipboard” it is then less of a horizontal shift than if it is from “Copy error details to clipboard” to “Copied to clipboard”. If that text is retained, then technically it should rather have the text “Copied error details to clipboard”.

The shorter text also means less horizontal width for a space that is already getting cramped.

johnwatkins0 replied:
I like this idea.

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -1619,6 +1643,13 @@ public static function filter_tag_row_actions( $actions, WP_Term $tag ) {
 			);
+
+			$actions['copy'] = sprintf(
+				'<button type="button" aria-label="%s" class="single-url-detail-copy button-link" data-error-json="%s">%s</button>',
+				esc_attr__( 'Copy to clipboard', 'amp' ),
+				esc_attr( self::get_error_details_json( $term ) ),
+				esc_html__( 'Copy to clipboard', 'amp' )
```
Is the `aria-label` actually needed here?

```suggestion
				'<button type="button" class="single-url-detail-copy button-link" data-error-json="%s">%s</button>',
				esc_attr( self::get_error_details_json( $term ) ),
				esc_html__( 'Copy to clipboard', 'amp' )
```

johnwatkins0 replied:
I was blindly following the other buttons on the screen, but you're right.

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -1591,6 +1591,30 @@ public static function add_admin_notices() {
+			}
+		}
+
+		$json['status'] = (bool) ( (int) $term->term_group & self::ACCEPTED_VALIDATION_ERROR_BIT_MASK ) ? __( 'Removed', 'amp' ) : __( 'Kept', 'amp' );
```
I don't think we should use translations here. Also, we should capture the reviewed state. So I suggest:

```suggestion
		$json['removed']  = (bool) ( (int) $term->term_group & self::ACCEPTED_VALIDATION_ERROR_BIT_MASK );
		$json['reviewed'] = (bool) ( (int) $term->term_group & self::ACKNOWLEDGED_VALIDATION_ERROR_BIT_MASK );
```

johnwatkins0 committed:
Make 'copy to clipboard' text consistent

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -1591,6 +1591,30 @@ public static function add_admin_notices() {
+
+		$json['status'] = (bool) ( (int) $term->term_group & self::ACCEPTED_VALIDATION_ERROR_BIT_MASK ) ? __( 'Removed', 'amp' ) : __( 'Kept', 'amp' );
+
+		return wp_json_encode( $json, JSON_PRETTY_PRINT );
```
Since the slashes aren't needed in this context (and they impair readability):

```suggestion
		return wp_json_encode( $json, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES );
```

westonruter replied:
I suppose this is not needed if the JSON is re-parsed and re-serialized. If that's the case, then `JSON_PRETTY_PRINT` isn't needed either.

johnwatkins0 replied:
Updated in c83dbb81901fdfb3d1edd6ba0fc626354e979ed7 and 54ecc46aa9b77af20658677cb5fbc6497d360b2f

westonruter commented on assets/src/amp-validation/copy-to-clipboard-buttons.js:
```diff
@@ -0,0 +1,84 @@
+	// eslint-disable-next-line no-new
+	clipboards.push( new Clipboard( 'button.single-url-detail-copy', {
+		text: ( btn ) => {
+			const json = JSON.parse( btn.getAttribute( 'data-error-json' ) );
+			const statusSelect = btn.closest( 'tr' ).querySelector( '.amp-validation-error-status' );
+			json.status = statusSelect.options[ statusSelect.selectedIndex ].text;
+
+			return JSON.stringify( json, null, '\t' );
```
I don't think we actually need to capture the `status` here because we already have the DB-_saved_ status in the `data-error-json` already. If we do include it, then it should be put into a separate field to not override `status` (or rather `removed` and `reviewed` per above). 

```suggestion
			return btn.getAttribute( 'data-error-json' );
```

westonruter commented on assets/src/amp-validation/copy-to-clipboard-buttons.js:
```diff
@@ -0,0 +1,84 @@
+					return null;
+				}
+
+				const json = JSON.parse( copyButton.getAttribute( 'data-error-json' ) );
+				const statusSelect = row.querySelector( '.amp-validation-error-status' );
+				json.status = statusSelect.options[ statusSelect.selectedIndex ].text;
+
+				return json;
```
See note above.

```suggestion
				return JSON.parse( copyButton.getAttribute( 'data-error-json' ) );
```

johnwatkins0 committed:
Clipboard-friendly errors: use status from DB instead of from UI

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Clipboard-friendly errors: convert status to removed bool and add reviewed field

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter commented:
@johnwatkins0 Do you want to rebase this onto `develop`? That should fix the build errors.

johnwatkins0 committed:
No need to pretty-print error details on the backend

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/5209-clipboard-friendly-errors

johnwatkins0 committed:
Parse and re-stringify JSON data to prettify

johnwatkins0 committed:
Remove unnecessary aria-label

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 commented:
@westonruter I've addressed all feedback and merged in the latest `develop`, but as you can see, there is still a composer issue in the actions building the zip files. I saw this in another PR earlier.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5209-clipboard-friendly-errors

* 'develop' of github.com:ampproject/amp-wp: (174 commits)
  Temporarily force of Composer v1 in CI (#5543)
  Fix typo in Optimizer readme (#5535)
  Include file.php separately from class-wp-upgrader.php (#5539)
  Add PhpIncludeInspection annotation
  Add PhpUnusedParameterInspection annotation
  Capture the original block render_callbacks to detect originating plugin
  Fix Gutenberg 9.2.1 compat due to gutenberg_current_parsed_block_tracking()
  Ensure full coverage for test_get_plugin_dir
  Skip Facebook tests earlier when external-http; update skip message
  Fix CSSList splicing and eliminate reliance on buggy replace/splice methods
  Add more removed rules that cause splice issue
  Add charset rule to allowed_at_rules_retained test data
  Ensure -moz-document rule is removed after replaced
  Skip FB test once testing external-http suite
  Add test for PluginRegistry::get_plugin_dir()
  Ensure css_list is sanitized after -moz-document transformation
  Transform -moz-document at-rule Firefox hack into supports hack
  Fix lint error
  Manually bootstrap WP CLI
  Skip test if Facebook oEmbed endpoint is down
  ...

westonruter committed:
Merge branch 'feature/5209-clipboard-friendly-errors' of github.com:ampproject/amp-wp into feature/5209-clipboard-friendly-errors

* 'feature/5209-clipboard-friendly-errors' of github.com:ampproject/amp-wp:
  Remove unnecessary aria-label
  Parse and re-stringify JSON data to prettify
  No need to pretty-print error details on the backend

westonruter commented:
> there is still a composer issue in the actions building the zip files. I saw this in another PR earlier.

Isn't this addressed by #5543?

westonruter reviewed (approved):
Love this functionality 👍

pierlon commented:
> > there is still a composer issue in the actions building the zip files. I saw this in another PR earlier.
> 
> Isn't this addressed by #5543?

@johnwatkins0 is referring to the current GHA workflow used to build the plugin; it's also now using composer v2 which is causing the jobs to fail.

johnwatkins0 commented:
> Isn't this addressed by #5543?

I believe that covered Travis only. I mentioned this to @pierlon on Slack today and he seems to have a handle on it as part of his GHA update.

westonruter committed:
Fix test_get_error_details_json

westonruter commented:
I've run the tests locally and they pass (aside from the oEmbed handlers that . So I'm going ahead and merging.
